src/libtrx/game/items/walkable.c

```diff
@@ -0,0 +1,167 @@
+#include "game/items/walkable.h"
+
+#include "game/game_buf.h"
+#include "game/items.h"
+#include "game/objects/general/drawbridge.h"
+#include "game/objects/general/trapdoor.h"
+#include "game/objects/vars.h"
+#include "game/rooms.h"
+#include "log.h"
+#include "memory.h"
+#include "vector.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+static SECTOR *M_GetItemPitSector(const ITEM *item, XYZ_32 pos);
+static void M_InsertSorted(WALKABLE **walkables, WALKABLE *node);
+void M_Add(int16_t item_num, const ITEM *item, XYZ_32 pos);
+
+static SECTOR *M_GetItemPitSector(const ITEM *const item, const XYZ_32 pos)
+{
+    int16_t room_num = item->room_num;
+    SECTOR *sector = Room_GetSector(pos.x, pos.y, pos.z, &room_num);
+    // LOG_DEBUG("room_num: %d", room_num);
+    return Room_GetPitSector(sector, pos.x, pos.z);
+}
+
+static void M_InsertSorted(WALKABLE **walkables, WALKABLE *node)
+{
+    while (*walkables && (*walkables)->pos.y >= node->pos.y) {
+        walkables = &(*walkables)->next;
+    }
+
+    node->next = *walkables;
+    *walkables = node;
+}
+
+void M_Add(const int16_t item_num, const ITEM *const item, const XYZ_32 pos)
+{
+    SECTOR *sector = M_GetItemPitSector(item, pos);
+
+    // Check if the walkable is already in the sector.
+    WALKABLE **walkables = &sector->walkable;
+    while (*walkables) {
+        if ((*walkables)->item_num == item_num) {
+            LOG_DEBUG("item_num: %d already exists in this sector", item_num);
+            return;
+        }
+        walkables = &(*walkables)->next;
+    }
+
+    LOG_DEBUG(
+        "Add %d; xyz: %d %d %d; floor.height: %d; ceiling.height: %d", item_num,
+        pos.x, pos.y, pos.z, sector->floor.height, sector->ceiling.height);
+    WALKABLE *node = GameBuf_Alloc(sizeof(WALKABLE), GBUF_WALKABLES);
```

Comment:
`const` and stuff, here in other pointers (I'd appreciate a general recheck prior to final merge – I'll avoid pinpointing every case to avoid noise in the CR)